pkg/controllers/interruption/controller.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2024 The CloudPilot AI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package interruption
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	karpv1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/events"
+
+	"github.com/cloudpilot-ai/karpenter-provider-alicloud/pkg/apis/v1alpha1"
+	"github.com/cloudpilot-ai/karpenter-provider-alicloud/pkg/cache"
+	interruptionevents "github.com/cloudpilot-ai/karpenter-provider-alicloud/pkg/controllers/interruption/events"
+)
+
+const (
+	ConditionTypeInstanceExpired = "InstanceExpired"
+)
+
+// Controller is an AliCloud interruption controller.
+type Controller struct {
+	kubeClient client.Client
+	recorder   events.Recorder
+
+	unavailableOfferingsCache *cache.UnavailableOfferings
+}
+
+func NewController(kubeClient client.Client, recorder events.Recorder,
+	unavailableOfferingsCache *cache.UnavailableOfferings) *Controller {
+	return &Controller{
+		kubeClient: kubeClient,
+		recorder:   recorder,
+
+		unavailableOfferingsCache: unavailableOfferingsCache,
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, node *corev1.Node) (reconcile.Result, error) {
+	nodeConditions := node.Status.Conditions
+
+	for ci := range nodeConditions {
+		if nodeConditions[ci].Type == ConditionTypeInstanceExpired && nodeConditions[ci].Status == corev1.ConditionTrue {
```

Comment:
Where is condition added, I didn’t find from karpenter-core.

Reply:
This is a notification provided by AlibabaCloud @daimaxiaxie 

https://help.aliyun.com/zh/ack/ack-managed-and-ack-dedicated/user-guide/best-practices-for-preemptible-instance-based-node-pools#section-lan-6cl-jq1

Reply:
Ok, I get it. If NPD components is not installed by default, prompt the user.